File: src/org/labrad/qubits/Experiment.java
package org.labrad.qubits;

import java.util.List;
import java.util.Map;
import java.util.Set;

import org.labrad.data.Data;
import org.labrad.qubits.channels.Channel;
import org.labrad.qubits.channels.PreampChannel;
import org.labrad.qubits.channels.TimingChannel;
import org.labrad.qubits.enums.DacTriggerId;
import org.labrad.qubits.mem.MemoryCommand;
import org.labrad.qubits.resources.AdcBoard;
import org.labrad.qubits.resources.AnalogBoard;
import org.labrad.qubits.resources.DacBoard;
import org.labrad.qubits.resources.MicrowaveBoard;

import com.google.common.base.Preconditions;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;


/**
 * "Experiment holds all the information about the fpga sequence as it is being built,
 * and knows how to produce the memory and sram instructions that actually get sent out to run the sequence."
 * 
 * For the ADC addition, we now have to be careful to only perform things like memory ops on DAC FpgaModels, not ADC ones.
 * 
 * @author maffoo
 * @author pomalley
 */
public class Experiment {

  /**
   * Create a new experiment using the given list of devices.
   * @param devices
   */
  public Experiment(List<Device> devices) {
    for (Device dev : devices) {
      addDevice(dev);
    }
    createResourceModels();
  }


  //
  // Resources
  //

  private void createResourceModels() {
    Map<DacBoard, FpgaModel> boards = Maps.newHashMap();

    // build models for all required resources
    for (Channel ch : getChannels()) {
      DacBoard board = ch.getDacBoard();
      FpgaModel fpga = boards.get(board);
      if (fpga == null) {
        if (board instanceof AnalogBoard) {
          fpga = new FpgaModelAnalog((AnalogBoard)board, this);
        } else if (board instanceof MicrowaveBoard) {
          fpga = new FpgaModelMicrowave((MicrowaveBoard)board, this);
        } else if (board instanceof AdcBoard) {
          fpga = new FpgaModelAdc((AdcBoard)board, this);
        } else {
          throw new RuntimeException("Unknown DAC board type for board " + board.getName());
        }
        boards.put(board, fpga);
        addFpga(fpga);
      }
      // connect this channel to the experiment and fpga model
      ch.setExperiment(this);
      ch.setFpgaModel(fpga);
    }

    // build lists of FPGA boards that have or don't have a timing channel
    nonTimerFpgas.addAll(getDacFpgas());
    for (PreampChannel ch : getChannels(PreampChannel.class)) {
      FpgaModelDac fpga = ch.getFpgaModel();
      timerFpgas.add(fpga);
      nonTimerFpgas.remove(fpga);
    }
  }


  //
  // Devices
  //

  private final List<Device> devices = Lists.newArrayList();
  private final Map<String, Device> devicesByName = Maps.newHashMap();

  private void addDevice(Device dev) {
    devices.add(dev);
    devicesByName.put(dev.getName(), dev);
  }

  public Device getDevice(String name) {
    Preconditions.checkArgument(devicesByName.containsKey(name),
        "Device '%s' not found.", name);
    return devicesByName.get(name);
  }

  private List<Device> getDevices() {
    return devices;
  }

  public List<Channel> getChannels() {
    return getChannels(Channel.class);
  }

  public <T extends Channel> List<T> getChannels(Class<T> cls) {
    List<T> channels = Lists.newArrayList();
    for (Device dev : devices) {
      channels.addAll(dev.getChannels(cls));
    }
    return channels;
  }


  //
  // FPGAs
  //

  private final Set<FpgaModel> fpgas = Sets.newHashSet();
  private final Set<FpgaModelDac> timerFpgas = Sets.newHashSet();
  private final Set<FpgaModelDac> nonTimerFpgas = Sets.newHashSet();

  private void addFpga(FpgaModel fpga) {
    fpgas.add(fpga);
  }

  /**
   * Get a list of FPGAs involved in this experiment
   */
  public Set<FpgaModel> getFpgas() {
    return Sets.newHashSet(fpgas);
  }

  public Set<FpgaModelDac> getTimerFpgas() {
    return Sets.newHashSet(timerFpgas);
  }

  public Set<FpgaModelDac> getNonTimerFpgas() {
    return Sets.newHashSet(nonTimerFpgas);
  }

  public Set<FpgaModelMicrowave> getMicrowaveFpgas() {
    Set<FpgaModelMicrowave> fpgas = Sets.newHashSet();
    for (FpgaModel fpga : this.fpgas) {
      if (fpga instanceof FpgaModelMicrowave) {
        fpgas.add((FpgaModelMicrowave)fpga);
      }
    }
    return fpgas;
  }
  
  /**
   * Many operations are only performed on DAC fpgas.
   * @return A set of all FpgaModelDac's in this experiment.
   * @author pomalley
   */
  
  public Set<FpgaModelDac> getDacFpgas() {
	  Set<FpgaModelDac> fpgas = Sets.newHashSet();
	  for (FpgaModel fpga : this.fpgas) {
		  if (fpga instanceof FpgaModelDac) {
			  fpgas.add((FpgaModelDac)fpga);
		  }
	  }
	  return fpgas;
  }
  
  /**
   * Conversely, sometimes we need the ADC fpgas. 
   * @return A set of all FpgaModelAdc's in this experiment.
   * @author pomalley
   */
  public Set<FpgaModelAdc> getAdcFpgas() {
	  Set<FpgaModelAdc> fpgas = Sets.newHashSet();
	  for (FpgaModel fpga : this.fpgas) {
		  if (fpga instanceof FpgaModelAdc) {
			  fpgas.add((FpgaModelAdc)fpga);
		  }
	  }
	  return fpgas;
  }

  public List<String> getFpgaNames() {
    List<String> boardsToRun = Lists.newArrayList();
    for (FpgaModel fpga : fpgas) {
      boardsToRun.add(fpga.getName());
    }
    return boardsToRun;
  }

  private final List<Data> setupPackets = Lists.newArrayList();
  private final List<String> setupState = Lists.newArrayList();
  private List<TimingChannel> timingOrder = null;
  private DacTriggerId autoTriggerId = null;
  private int autoTriggerLen = 0;

  /**
   * Clear all configuration that has been set for this experiment
   */
  public void clearConfig() {
    // reset setup packets
    clearSetupState();

    // clear timing order
    timingOrder = null;

    // clear autotrigger
    autoTriggerId = null;

    // clear configuration on all channels
    for (Device dev : getDevices()) {
      for (Channel ch : dev.getChannels()) {
        ch.clearConfig();
      }
    }
  }


  private void clearSetupState() {
    setupState.clear();
    setupPackets.clear();
  }

  public void setSetupState(List<String> state, List<Data> packets) {
    clearSetupState();
    setupState.addAll(state);
    setupPackets.addAll(packets);
  }

  public List<String> getSetupState() {
    return Lists.newArrayList(setupState);
  }

  public List<Data> getSetupPackets() {
    return Lists.newArrayList(setupPackets);
  }

  public void setAutoTrigger(DacTriggerId id, int length) {
    autoTriggerId = id;
    autoTriggerLen = length;
  }

  public DacTriggerId getAutoTriggerId() {
    return autoTriggerId;
  }

  public int getAutoTriggerLen() {
    return autoTriggerLen;
  }

  public void setTimingOrder(List<TimingChannel> channels) {
    timingOrder = Lists.newArrayList(channels);
  }

  /**
   * Get the order of boards from which to return timing data
   * @return
   */
  public List<String> getTimingOrder() {
    List<String> order = Lists.newArrayList();
    for (TimingChannel ch : getTimingChannels()) {
      order.add(ch.getDacBoard().getName());
    }
    return order;
  }

  public List<TimingChannel> getTimingChannels() {
    return timingOrder != null ? timingOrder : getChannels(TimingChannel.class);
  }

  //
  // Memory
  //

  /**
   * Clear the memory content for this experiment
   * 
   * This only applies to DAC fpgas.
   */
  public void clearMemory() {
    // all memory state is kept in the fpga models, so we clear them out
    for (FpgaModelDac fpga : getDacFpgas()) {
      fpga.clearMemory();
    }
  }

  /**
   * Add bias commands to a set of FPGA boards. Only applies to DACs.
   * @param allCmds
   */
  public void addBiasCommands(ListMultimap<FpgaModelDac, MemoryCommand> allCmds, double delay) {
    // find the maximum number of commands on any single fpga board
    int maxCmds = 0;
    for (FpgaModelDac fpga : allCmds.keySet()) {
      maxCmds = Math.max(maxCmds, allCmds.get(fpga).size());
    }

    // add commands for each board, including noop padding and final delay
    for (FpgaModelDac fpga : getDacFpgas()) {
      List<MemoryCommand> cmds = allCmds.get(fpga); 
      if (cmds != null) {
        fpga.addMemoryCommands(cmds);
        fpga.addMemoryNoops(maxCmds - cmds.size());
      } else {
        fpga.addMemoryNoops(maxCmds);
      }
      if (delay > 0) {
        fpga.addMemoryDelay(delay);
      }
    }
  }

  /**
   * Add a delay in the memory sequence.
   * Only applies to DACs.
   */
  public void addMemoryDelay(double microseconds) {
    for (FpgaModelDac fpga : getDacFpgas()) {
      fpga.addMemoryDelay(microseconds);
    }
  }

  /**
   * Call SRAM. Only applies to DACs.
   */
  public void callSramBlock(String block) {
    for (FpgaModelDac fpga : getDacFpgas()) {
      fpga.callSramBlock(block);
    }
  }

  public void callSramDualBlock(String block1, String block2) {
    for (FpgaModelDac fpga : getDacFpgas()) {
      fpga.callSramDualBlock(block1, block2);
    }
  }

  public void setSramDualBlockDelay(double delay) {
    for (FpgaModelDac fpga : getDacFpgas()) {
      fpga.setSramDualBlockDelay(delay);
    }
  }

  /**
   * Start timer on a set of boards.
   * This only applies to DAC fpgas.
   */
  public void startTimer(List<PreampChannel> channels) {
    Set<FpgaModelDac> starts = Sets.newHashSet();
    Set<FpgaModelDac> noops = getTimerFpgas();
    for (PreampChannel ch : channels) {
      FpgaModelDac fpga = ch.getFpgaModel();
      starts.add(fpga);
      noops.remove(fpga);
    }
    // non-timer boards get started if they have never been started before
    for (FpgaModelDac fpga : getNonTimerFpgas()) {
      if (!fpga.isTimerStarted()) {
        starts.add(fpga);
      } else {
        noops.add(fpga);
      }
    }
    // start the timer on requested boards
    for (FpgaModelDac fpga : starts) {
      fpga.startTimer();
    }
    // insert a no-op on all other boards
    for (FpgaModelDac fpga : noops) {
      fpga.addMemoryNoop();
    }
  }

  /**
   * Stop timer on a set of boards.
   */
  public void stopTimer(List<PreampChannel> channels) {
    Set<FpgaModelDac> stops = Sets.newHashSet();
    Set<FpgaModelDac> noops = getTimerFpgas();
    for (PreampChannel ch : channels) {
      FpgaModelDac fpga = ch.getFpgaModel();
      stops.add(fpga);
      noops.remove(fpga);
    }
    // stop non-timer boards if they are currently running
    for (FpgaModelDac fpga : getNonTimerFpgas()) {
      if (fpga.isTimerRunning()) {
        stops.add(fpga);
      } else {
        noops.add(fpga);
      }
    }
    // stop the timer on requested boards and non-timer boards
    for (FpgaModelDac fpga : stops) {
      fpga.stopTimer();
    }
    // insert a no-op on all other boards
    for (FpgaModelDac fpga : noops) {
      fpga.addMemoryNoop();
    }
  }


  //
  // SRAM
  //


  private List<String> blocks = Lists.newArrayList();
  private Map<String, Integer> blockLengths = Maps.newHashMap();

  public void startSramBlock(String name, long length) {
    blocks.add(name);
    blockLengths.put(name, (int)length);
  }

  public List<String> getBlockNames() {
    return Lists.newArrayList(blocks);
  }

  public int getBlockLength(String name) {
    Preconditions.checkArgument(blockLengths.containsKey(name), "SRAM block '%s' is undefined", name);
    return blockLengths.get(name);
  }

  /**
   * Get the proper length for an SRAM block after padding.
   * The length should be a multiple of 4 and greater or equal to 20.
   * @param name
   * @return
   */
  public int getPaddedBlockLength(String name) {
    int len = getBlockLength(name);
    if (len % 4 == 0 && len >= 20) return len;
    int paddedLen = Math.max(len + ((4 - len % 4) % 4), 20);
    return paddedLen;
  }

  public int getBlockStartAddress(String name) {
    int start = 0;
    for (String block : blocks) {
      if (block.equals(name)) {
        return start;
      }
      start += getPaddedBlockLength(block);
    }
    throw new RuntimeException(String.format("Block '%s' not found", name));
  }

  public int getBlockEndAddress(String name) {
    int end = 0;
    for (String block : blocks) {
      end += getPaddedBlockLength(block);
      if (block.equals(name)) {
        return end - 1;
      }
    }
    throw new RuntimeException(String.format("Block '%s' not found", name));
  }
}

File: QubitServer/src/main/scala/org/labrad/qubits/Experiment.scala
package org.labrad.qubits

import org.labrad.data.Data
import org.labrad.qubits.channels._
import org.labrad.qubits.enums.DacTriggerId
import org.labrad.qubits.mem.MemoryCommand
import org.labrad.qubits.resources.AdcBoard
import org.labrad.qubits.resources.AnalogBoard
import org.labrad.qubits.resources.DacBoard
import org.labrad.qubits.resources.MicrowaveBoard
import org.labrad.types._
import scala.collection.mutable
import scala.reflect.ClassTag

/**
 * "Experiment holds all the information about the fpga sequence as it is being built,
 * and knows how to produce the memory and sram instructions that actually get sent out to run the sequence."
 *
 * For the ADC addition, we now have to be careful to only perform things like memory ops on DAC FpgaModels, not ADC ones.
 *
 * @author maffoo
 * @author pomalley
 */
class Experiment(val devices: Seq[Device]) {

  private val devicesByName = devices.map { dev =>
    dev.name -> dev
  }.toMap

  private val boards = mutable.Map.empty[DacBoard, FpgaModel]

  // build models for all required resources
  for (ch <- channels[FpgaChannel]) {
    val board = ch.dacBoard
    val fpga = boards.getOrElseUpdate(board, {
      board match {
        case board: AnalogBoard => new FpgaModelAnalog(board, this)
        case board: MicrowaveBoard => new FpgaModelMicrowave(board, this)
        case board: AdcBoard => new FpgaModelAdc(board, this)
        case _ => sys.error(s"Unknown DAC board type for board ${board.name}")
      }
    })
    // connect this channel to the correct fpga model
    ch.setFpgaModel(fpga)
  }

  for (ch <- channels[FastBiasSerialChannel]) {
    // TODO: how to represent DC rack hardware in the experiment?
  }

  val fpgas = boards.values.toSet

  // sets of FPGA boards of various types
  val microwaveFpgas = fpgas.collect { case fpga: FpgaModelMicrowave => fpga }
  val dacFpgas = fpgas.collect { case dac: FpgaModelDac => dac }
  val adcFpgas = fpgas.collect { case adc: FpgaModelAdc => adc }

  // build sets of FPGA boards that have or don't have a timing channel
  val timerFpgas = channels[PreampChannel].map(_.fpgaModel).toSet
  val nonTimerFpgas = dacFpgas -- timerFpgas

  private val _setupPackets = mutable.Buffer.empty[Data]
  private val _setupState = mutable.Buffer.empty[String]
  private var _timingOrder: Seq[TimingOrderItem] = null
  private var _autoTriggerId: DacTriggerId = null
  private var _autoTriggerLen = 0

  private var _loopDelay: Double = 0
  private var _loopDelayConfigured = false

  //
  // Devices
  //

  def device(name: String): Device = {
    devicesByName.get(name).getOrElse {
      sys.error(s"Device '$name' not found.")
    }
  }

  def channels[T <: Channel : ClassTag]: Seq[T] = {
    devices.flatMap(_.getChannels[T])
  }


  //
  // FPGAs
  //

  /**
   * Clear the memory or jump table commands for these FPGAs.
   */
  def clearControllers(): Unit = {
    for (fpga <- dacFpgas) {
      fpga.clearController()
    }
  }

  def fpgaNames: Seq[String] = {
    fpgas.map(_.name).toSeq
  }

  /**
   * Clear all configuration that has been set for this experiment
   */
  def clearConfig(): Unit = {
    // reset setup packets
    clearSetupState()

    // clear timing order
    _timingOrder = null

    // clear autotrigger
    _autoTriggerId = null

    // clear configuration on all channels
    for {
      dev <- devices
      ch <- dev.getChannels()
    } {
      ch.clearConfig()
    }

    // de-configure loopDelay
    _loopDelayConfigured = false
  }


  private def clearSetupState(): Unit = {
    _setupState.clear()
    _setupPackets.clear()
  }

  def setSetupState(state: Seq[String], packets: Seq[Data]): Unit = {
    clearSetupState()
    _setupState ++= state
    _setupPackets ++= packets
  }

  def setupState: Seq[String] = {
    _setupState.toVector
  }

  def setupPackets: Seq[Data] = {
    _setupPackets.toVector
  }

  def setAutoTrigger(id: DacTriggerId, length: Int): Unit = {
    _autoTriggerId = id
    _autoTriggerLen = length
  }

  def autoTriggerId: DacTriggerId = {
    _autoTriggerId
  }

  def autoTriggerLen: Int = {
    _autoTriggerLen
  }

  def setTimingOrder(to: Seq[TimingOrderItem]): Unit = {
    _timingOrder = to
  }

  def configLoopDelay(loopDelay: Double): Unit = {
    _loopDelay = loopDelay
    _loopDelayConfigured = true
  }

  def isLoopDelayConfigured(): Boolean = {
    _loopDelayConfigured
  }
  def loopDelay: Double = {
    _loopDelay
  }

  /**
   * Get the order of boards from which to return timing data
   * @return
   */
  def timingOrder: Seq[String] = {
    timingChannels.map(_.toString)
  }

  def timingChannels: Seq[TimingOrderItem] = {
    // if we have an existing timing order, use it
    if (_timingOrder != null) {
      _timingOrder
    // if not, use everything--all DACs, all ADCs/active ADC channels
    } else {
      channels[TimingChannel].map {
        case t: AdcChannel => new TimingOrderItem(t, t.demodChannel)
        case t => new TimingOrderItem(t)
      }
    }
  }

  def adcTimingOrderIndices(): Seq[Int] = {
    timingChannels.zipWithIndex.collect {
      case (toi, i) if toi.isAdc => i
    }
  }

  def dacTimingOrderIndices(): Seq[Int] = {
    timingChannels.zipWithIndex.collect {
      case (toi, i) if !toi.isAdc => i
    }
  }

  //
  // Jump Table
  //

  def addJumpTableEntry(commandName: String, commandData: Data): Unit = {
    for (fpga <- dacFpgas) {
      fpga.jumpTableController.addJumpTableEntry(commandName, commandData)
    }
  }


  //
  // Memory
  //

  /**
   * Add bias commands to a set of FPGA boards. Only applies to DACs.
   * @param allCmds
   */
  def addBiasCommands(allCmds: Map[FpgaModelDac, Seq[MemoryCommand]], delay: Double): Unit = {
    // find the maximum number of commands on any single fpga board
    var maxCmds = 0
    for ((fpga, cmds) <- allCmds) {
      maxCmds = Math.max(maxCmds, cmds.size)
    }

    // add commands for each board, including noop padding and final delay
    for (fpga <- dacFpgas) {
      allCmds.get(fpga) match {
        case Some(cmds) =>
          fpga.memoryController.addMemoryCommands(cmds)
          fpga.memoryController.addMemoryNoops(maxCmds - cmds.size)

        case None =>
          fpga.memoryController.addMemoryNoops(maxCmds)
      }
      if (delay > 0) {
        fpga.memoryController.addMemoryDelay(delay)
      }
    }
  }

  /**
   * Add a delay command to exactly one board
   *
   */
  def addSingleMemoryDelay(fpga: FpgaModelDac, delay_us: Double): Unit = {
    fpga.memoryController.addMemoryDelay(delay_us)
  }

  /**
   * Add a delay in the memory sequence of all boards.
   * Only applies to DACs.
   */
  def addMemoryDelay(microseconds: Double): Unit = {
    for (fpga <- dacFpgas) {
      fpga.memoryController.addMemoryDelay(microseconds)
    }
  }

  def addMemSyncDelay(): Unit = {
    //Find maximum sequence length on all fpgas
    var maxT_us = 0.0
    for (fpga <- fpgas) {
      try {
        val t_us = fpga.sequenceLengthPostSRAM_us
        maxT_us = Math.max(maxT_us, t_us)
      } catch {
        case ex: java.lang.IllegalArgumentException =>
      }
    }

    for (fpga <- dacFpgas) {
      var t = 0.0
      try {
        t = fpga.sequenceLength_us
      } catch {
        case ex: java.lang.IllegalArgumentException =>
      }
      if (t < maxT_us) {
        fpga.memoryController.addMemoryDelay(maxT_us - t)
      } else {
        fpga.memoryController.addMemoryNoop()
      }
    }
  }

  /**
   * Call SRAM. Only applies to DACs.
   */
  def callSramBlock(block: String): Unit = {
    for (fpga <- dacFpgas) {
      fpga.memoryController.callSramBlock(block)
    }
  }

  def callSramDualBlock(block1: String, block2: String): Unit = {
    for (fpga <- dacFpgas) {
      fpga.memoryController.callSramDualBlock(block1, block2)
    }
  }

  def setSramDualBlockDelay(delay_ns: Double): Unit = {
    for (fpga <- dacFpgas) {
      fpga.memoryController.setSramDualBlockDelay(delay_ns)
    }
  }

  /**
   * Get the length of the shortest SRAM block across all fpgas.
   * @return
   */
  def shortestSram: Int = {
    val lens = for {
      fpga <- dacFpgas
      block <- fpga.blockNames
    } yield fpga.blockLength(block)

    if (lens.isEmpty) 0 else lens.min
  }

  /**
   * Start timer on a set of boards.
   * This only applies to DAC fpgas.
   */
  def startTimer(channels: Seq[PreampChannel]): Unit = {
    val boards = channels.map(_.fpgaModel).toSet

    // start requested timers
    val timerStarts = boards
    val timerNoops = timerFpgas -- boards

    // start non timer boards that have never been started
    val (nonTimerNoops, nonTimerStarts) = nonTimerFpgas.partition(_.memoryController.isTimerStarted)

    // start the timer on requested boards
    for (fpga <- timerStarts ++ nonTimerStarts) {
      fpga.memoryController.startTimer()
    }
    // insert a no-op on all other boards
    for (fpga <- timerNoops ++ nonTimerNoops) {
      fpga.memoryController.addMemoryNoop()
    }
  }

  /**
   * Stop timer on a set of boards.
   */
  def stopTimer(channels: Seq[PreampChannel]): Unit = {
    val boards = channels.map(_.fpgaModel).toSet

    // stop requested timers
    val timerStops = boards
    val timerNoops = timerFpgas -- boards

    // stop non-timer boards if they are currently running
    val (nonTimerStops, nonTimerNoops) = nonTimerFpgas.partition(_.memoryController.isTimerRunning)

    // stop the timer on requested boards and non-timer boards
    for (fpga <- timerStops ++ nonTimerStops) {
      fpga.memoryController.stopTimer()
    }
    // insert a no-op on all other boards
    for (fpga <- timerNoops ++ nonTimerNoops) {
      fpga.memoryController.addMemoryNoop()
    }
  }

}

// stupid handler class to implement a timing order item
class TimingOrderItem(val channel: TimingChannel, subChannel: Int = -1) {

  override def toString(): String = {
    if (subChannel == -1)
      channel.dacBoard.name
    else
      channel.dacBoard.name + "::" + subChannel
  }

  def isAdc(): Boolean = {
    channel.isInstanceOf[AdcChannel]
  }

  /**
   * @param data Must be *w (DACs) or (*i{I}, *i{Q}) (ADCs)
   * @return T/F for 1/0 qubit state for each item in data.
   */
  def interpretData(data: Data): Array[Boolean] = {
    channel match {
      case adc: AdcChannel =>
        require(data.t == Type("(*i, *i)"),
            s"interpretData called with data type ${data.t} on an ADC channel. Qubit Sequencer mixup.")
        val (is, qs) = data.get[(Array[Int], Array[Int])]
        adc.interpretPhases(is, qs)

      case preamp: PreampChannel =>
        require(data.t == Type("*w"),
            s"interpretData called with data type ${data.t} on a DAC channel. Qubit Sequencer mixup.")
        preamp.interpretSwitches(data.get[Array[Long]])
    }
  }
}
